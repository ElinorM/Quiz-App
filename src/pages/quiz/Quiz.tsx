/** @jsxImportSource @emotion/react */
import React, { useState } from 'react';
import { QuestionPage } from './components/questionPage/QuestionPage';
import { QUESTIONS } from './quiz.constants';
import { Question } from './quiz.types';
import { ScorePage } from './components/scorePage/ScorePage';
import { QuizButtons } from './components/quizButtons/QuizButtons';
import { css } from '@emotion/react';

const style = {
  wrapper: css({
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center'
  })
};
export const Quiz = () => {
  const [questions, setQuestions] = useState(QUESTIONS as Question[]);
  const [questionCount, setQuestionCount] = useState(0);
  const [submitted, setSubmitted] = useState(false);

  const updateSelectedOption = (selectedOption: string) => {
    const updatedQuestions = [...questions];
    if(questions[questionCount].selectedOption.includes(selectedOption)) {
      updatedQuestions[questionCount].selectedOption = updatedQuestions[questionCount].selectedOption.filter(option => option !== selectedOption);
      setQuestions(updatedQuestions);
      return;
    }
    updatedQuestions[questionCount].selectedOption.push(selectedOption);
    setQuestions(updatedQuestions);
  }

  return (
    <div css={style.wrapper}>
      <h1>Quiz App</h1>
      {submitted
        ? <ScorePage questions={questions}/>
        : <div>
            <QuestionPage question={questions[questionCount]} updateSelectedOption={updateSelectedOption}/>
            <QuizButtons question={questions[questionCount]} questionCount={questionCount} setQuestionCount={setQuestionCount} setSubmitted={setSubmitted}/>
        </div>
      }
    </div>
  );
}
