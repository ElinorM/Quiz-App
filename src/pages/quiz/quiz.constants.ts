export const QUESTIONS = [
  {
    title: "What is the capital of France?",
    options: ["New York", "London", "Paris", "Dublin"],
    answer: ["Paris", 'London'],
    selectedOption: []
  },
  {
    title: "What is the capital of India?",
    options: ["New York", "London", "Paris", "Delhi"],
    answer: ["Delhi"],
    selectedOption: []
  },
  {
    title: "What is the capital of USA?",
    options: ["New York", "Washington DC", "Paris", "Dublin"],
    answer: ["Washington DC"],
    selectedOption: []
  },
  {
    title: "What is the capital of Australia?",
    options: ["New York", "Canberra", "Paris", "Dublin"],
    answer: ["Canberra"],
    selectedOption: []
  },
  {
    title: "What is the capital of Japan?",
    options: ["New York", "London", "Tokyo", "Dublin"],
    answer: ["Tokyo"],
    selectedOption: []
  },
]
