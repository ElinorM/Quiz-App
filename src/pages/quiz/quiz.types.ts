export type Question = {
  title: string;
  options: string[];
  answer: string[];
  selectedOption: string[];
}
