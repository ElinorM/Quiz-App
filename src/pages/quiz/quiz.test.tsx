import React from 'react';
import { render, screen, } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Quiz } from './Quiz';
import { QUESTIONS } from './quiz.constants';


describe('Quiz', () => {
  it('should match snapshot', () => {
    expect(render(<Quiz/>)).toMatchSnapshot();
  });

  it('should render the first question', () => {
    render(<Quiz/>);
    expect(screen.getByText(QUESTIONS[0].title)).toBeInTheDocument();
  });

  it('should show the next question when Next is clicked', async () => {
    render(<Quiz/>);
    await userEvent.click(screen.getByRole('checkbox', { name: 'London' }));
    await userEvent.click(screen.getByRole('button', { name: 'Next' }));
    expect(screen.getByText(QUESTIONS[1].title)).toBeInTheDocument();
  });

  it('should display the ScorePage when the quiz is submitted', async () => {
    render(<Quiz/>);
    for (let i = 0; i < QUESTIONS.length - 1; i++) {
      await userEvent.click(screen.getByRole('checkbox', { name: QUESTIONS[i].options[2] }));
      await userEvent.click(screen.getByRole('button', { name: 'Next' }));
    }
    await userEvent.click(screen.getByRole('checkbox', { name: QUESTIONS[QUESTIONS.length - 1].options[2] }));
    await userEvent.click(screen.getByRole('button', { name: 'Submit' }));
    expect(screen.getByText('Score:', { exact: false })).toBeInTheDocument();
  });
});
