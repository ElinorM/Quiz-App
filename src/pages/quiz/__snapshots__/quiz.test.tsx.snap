// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="css-dvxtzn"
      >
        <h1>
          Quiz App
        </h1>
        <div>
          <div>
            <h2>
              What is the capital of France?
            </h2>
            <ul>
              <div>
                <label>
                  <input
                    id="0"
                    name="New York"
                    type="checkbox"
                    value="New York"
                  />
                  New York
                </label>
              </div>
              <div>
                <label>
                  <input
                    id="1"
                    name="London"
                    type="checkbox"
                    value="London"
                  />
                  London
                </label>
              </div>
              <div>
                <label>
                  <input
                    id="2"
                    name="Paris"
                    type="checkbox"
                    value="Paris"
                  />
                  Paris
                </label>
              </div>
              <div>
                <label>
                  <input
                    id="3"
                    name="Dublin"
                    type="checkbox"
                    value="Dublin"
                  />
                  Dublin
                </label>
              </div>
            </ul>
          </div>
          <div
            class="css-1603fsx"
          >
            <button
              disabled=""
            >
              Next
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-dvxtzn"
    >
      <h1>
        Quiz App
      </h1>
      <div>
        <div>
          <h2>
            What is the capital of France?
          </h2>
          <ul>
            <div>
              <label>
                <input
                  id="0"
                  name="New York"
                  type="checkbox"
                  value="New York"
                />
                New York
              </label>
            </div>
            <div>
              <label>
                <input
                  id="1"
                  name="London"
                  type="checkbox"
                  value="London"
                />
                London
              </label>
            </div>
            <div>
              <label>
                <input
                  id="2"
                  name="Paris"
                  type="checkbox"
                  value="Paris"
                />
                Paris
              </label>
            </div>
            <div>
              <label>
                <input
                  id="3"
                  name="Dublin"
                  type="checkbox"
                  value="Dublin"
                />
                Dublin
              </label>
            </div>
          </ul>
        </div>
        <div
          class="css-1603fsx"
        >
          <button
            disabled=""
          >
            Next
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
